remote_attestation/rust/src/message.rs

```diff
@@ -0,0 +1,251 @@
+    pub ephemeral_public_key: [u8; EPHEMERAL_PUBLIC_KEY_SIZE_BYTES],
+    /// Random vector sent in messages for preventing replay attacks.
+    pub random: [u8; REPLAY_PROTECTION_ARRAY_SIZE_BYTES],
+    /// Signature of the SHA-256 hash of all previously send and received messaged.
+    /// Transcript signature is sent in messages to prevent replay attacks.
+    ///
+    /// Signature must be an IEEE-P1363 encoded ECDSA-P256 signature.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    /// https://standards.ieee.org/standard/1363-2000.html
+    #[serde(with = "BigArray")]
+    pub transcript_signature: [u8; TRANSCRIPT_SIGNATURE_SIZE_BYTES],
+    /// Public key used to sign transcripts.
+    ///
+    /// Public key must be an OpenSSL ECDSA-P256 key, which is represented as
+    /// `0x04 | X: 32-byte | Y: 32-byte`.
+    /// Where X and Y are big-endian coordinates of an Elliptic Curve point.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    #[serde(with = "BigArray")]
+    pub signing_public_key: [u8; SIGNING_PUBLIC_KEY_SIZE_BYTES],
+    /// Information used for remote attestation such as a TEE report and a TEE provider's
+    /// certificate. TEE report contains a hash of the `signing_public_key`.
+    ///
+    /// Attestation info must be a serialized `oak.remote_attestation.AttestationInfo` Protobuf
+    /// message.
+    pub attestation_info: Vec<u8>,
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct ClientIdentity {
+    /// Message header.
+    header: u8,
+    /// Public key needed to establish a session key.
+    pub ephemeral_public_key: [u8; EPHEMERAL_PUBLIC_KEY_SIZE_BYTES],
+    /// Signature of the SHA-256 hash of all previously send and received messaged.
+    /// Transcript signature is sent in messages to prevent replay attacks.
+    ///
+    /// Signature must be an IEEE-P1363 encoded ECDSA-P256 signature.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    /// https://standards.ieee.org/standard/1363-2000.html
+    #[serde(with = "BigArray")]
+    pub transcript_signature: [u8; TRANSCRIPT_SIGNATURE_SIZE_BYTES],
+    /// Public key used to sign transcripts.
+    ///
+    /// Public key must be an OpenSSL ECDSA-P256 key, which is represented as
+    /// `0x04 | X: 32-byte | Y: 32-byte`.
+    /// Where X and Y are big-endian coordinates of an Elliptic Curve point.
+    /// https://datatracker.ietf.org/doc/html/rfc6979
+    #[serde(with = "BigArray")]
+    pub signing_public_key: [u8; SIGNING_PUBLIC_KEY_SIZE_BYTES],
+    /// Information used for remote attestation such as a TEE report and a TEE provider's
+    /// certificate. TEE report contains a hash of the `signing_public_key`.
+    ///
+    /// Attestation info must be a serialized `oak.remote_attestation.AttestationInfo` Protobuf
+    /// message.
+    pub attestation_info: Vec<u8>,
+}
+
+pub trait Serializable {
+    fn serialize(&self) -> anyhow::Result<Vec<u8>>;
+    fn deserialize(bytes: &[u8]) -> anyhow::Result<Self> where Self: Sized;
```

Comment:
This sounds more like a method in a `Deserializable` trait. Can the trait be split into two?

Reply:
Done